cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java

```diff
@@ -180,49 +178,68 @@ public boolean isSupported(ProgramType programType, CConfiguration cConf) {
-          throw Throwables.propagate(e);
-        }
-      }
+    return injector.createChildInjector(
+        new AbstractModule() {
+          @Override
+          protected void configure() {
+            // Add a binding for SparkCompat enum
+            // The binding needs to be on the enum class loaded by the runner classloader
+            try {
+              Class<? extends Enum> type =
+                  (Class<? extends Enum>) runnerClassLoader.loadClass(SparkCompat.class.getName());
+              bindEnum(binder(), type, providerSparkCompat.name());
+            } catch (ClassNotFoundException e) {
+              // This shouldn't happen
+              throw Throwables.propagate(e);
+            }
+          }
 
-      private <T extends Enum<T>> void bindEnum(Binder binder, Class<T> enumType, String name) {
-        // This workaround the enum generic
-        binder.bind(enumType).toInstance(Enum.valueOf(enumType, name));
-      }
-    });
+          private <T extends Enum<T>> void bindEnum(Binder binder, Class<T> enumType, String name) {
+            // This workaround the enum generic
+            binder.bind(enumType).toInstance(Enum.valueOf(enumType, name));
+          }
+        });
   }
 
-  private synchronized ClassLoader getDistributedRunnerClassLoader(boolean rewriteCheckpointTempFileName) {
-    try {
-      if (distributedRunnerClassLoader == null) {
-        // Never needs to rewrite yarn client in CDAP master, which is the only place using distributed program runner
-        distributedRunnerClassLoader = createClassLoader(true, false, rewriteCheckpointTempFileName);
+  /**
+   * Returns the singleton {@link ClassLoader} for distributed program execution. The distributed
+   * program runner is used by the CDAP master to launch program that execute remotely, hence
+   * it doesn't need to do any class rewrite. We only create the SparkRunnerClassLoader once and
+   * keep reusing it since in the CDAP master, there is no SparkContext being created, hence no need
+   * to provide runtime isolation. This also limits the amount of permgen usage to be constant in
+   * the CDAP master regardless of how many Spark programs are running. We never need to close the
+   * SparkRunnerClassLoader until process shutdown.
+   */
+  private ClassLoader getDistributedRunnerClassLoader() {
+    ClassLoader classLoader = distributedRunnerClassLoader;
+    if (classLoader != null) {
+      return classLoader;
+    }
+
+    synchronized (this) {
+      classLoader = distributedRunnerClassLoader;
+      if (classLoader != null) {
+        return classLoader;
+      }
+      try {
+        // Never needs to rewrite yarn client in CDAP master, which is the only place using
+        // distributed program runner. Also wrap the classloader so that it can't be closed.
+        distributedRunnerClassLoader = new ClassLoader(createClassLoader(true, false, false)) {};
+        return distributedRunnerClassLoader;
```

Comment:
Instead of returning `distributedRunnerClassLoader` and incurring another expensive read from a volatile variable, can we write the new classloader to our existing local variable `classLoader`, assign it to `distributedRunnerClassLoader`, then return the local variable?

Reply:
That's not really expensive and it is just have it hit once due to the singleton. Just don't wanna clutter the code too much